src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -421,37 +456,42 @@ protected long fetchOffsetTokenWithRetry() {
   }
 
   /**
-   * Fallback function to execute when retries for fetching getOffsetToken for channel fails.
+   * If a valid offset is found from snowflake, we will reset the topicPartition with
+   * (offsetReturnedFromSnowflake + 1).
+   *
+   * <p>Please note, this is done only after a failure in fetching offset. Idea behind resetting
+   * offset (1 more than what we found in snowflake) is that Kafka should send offsets from this
+   * offset number so as to not miss any data.
    *
-   * <p>Fallback function is to re-open the channel and fetch the latestOffsetToken one more time
-   * after reopen is successful.
+   * @return offset which was last present in Snowflake
+   */
+  private long offsetTokenFallbackSupplier() {
+    final long offsetRecoveredFromSnowflake = getRecoveredOffsetFromSnowflake();
```

Comment:
getRecovered offset from snowflake gives us what is present in snowflake. if nothing then -1. (It starts from 0)

So we return +1 of what we get from snowflake. 

We return -1 to original caller set to 0 in kafka since that is what we ask them to send from..